mllib/src/main/scala/org/apache/spark/ml/util/SchemaUtils.scala

```diff
@@ -106,6 +107,93 @@ private[spark] object SchemaUtils {
     StructType(schema.fields :+ col)
   }
 
+  /**
+   * Update the size of a ML Vector column. If this column do not exist, append it.
+   * @param schema input schema
+   * @param colName column name
+   * @param size number of features
+   * @return new schema
+   */
+  def updateAttributeGroupSize(
+      schema: StructType,
+      colName: String,
+      size: Int): StructType = {
+    require(size > 0)
+    val attrGroup = new AttributeGroup(colName, size)
+    val field = attrGroup.toStructField
+    updateField(schema, field, true)
+  }
+
+  /**
+   * Update the number of values of an existing column. If this column do not exist, append it.
+   * @param schema input schema
+   * @param colName column name
+   * @param numValues number of values.
+   * @return new schema
+   */
+  def updateNumValues(
+      schema: StructType,
+      colName: String,
+      numValues: Int): StructType = {
+    val attr = if (numValues == 2) {
+      BinaryAttribute.defaultAttr
+        .withName(colName)
+    } else {
+      NominalAttribute.defaultAttr
+        .withName(colName)
+        .withNumValues(numValues)
+    }
```

Comment:
Not sure about this. Is numValues == 2 case always BinaryAttribute? NominalAttribute can not have two number of values?

Reply:
Good point.
I found that existing impls like `Bucketizer` will check whether `numValues==2`,
so I think it is safe to only use `NominalAttribute` here.